### PR TITLE
SIMS #137 - Secure API end point with KeyCloak Token Verification

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
Enabled the JWT "offline" validation on the API using the passport package. After a team discussion, we decided to not validate the received token on API on every request using a "bearer-only" client on the Keyclock.

Implemented global authentication and roles-based authentication, as described on Nestjs docs.
https://docs.nestjs.com/security/authentication
https://docs.nestjs.com/security/authorization

Implemented the @UserToken parameter decorator to provided typed access to user token information, as shown below.
image

Retrieving the public key for JWT validation direct from the Keycloak issuer URL (e.g. https://dev.oidc.gov.bc.ca/auth/realms/jxoe2o46).

The public key is gonna be retrieved and converted to a PEM format to later be consumed by passport JWT validation.
More about PEM format: https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail

Replaced the nest-keycloak-connect by the packages based os passport.
Removed env variables KEYCLOAK_CLIENT_API  and KEYCLOAK_CLIENT_SECRET that are no longer needed.

Moved the config endpoint alongside the other controllers.

All the endpoints will require an authentication token unless they are decorated with the "Public" decorator, as shown below.
image

The roles authorization is also ready to be used as shown below.
image

### CODE REVIEW

File: sources/packages/api/src/app.module.ts [hunk 1]
```diff
@@ -2,18 +2,26 @@ import { Module } from "@nestjs/common";
 import { AppController } from "./app.controller";
 import { AppService } from "./app.service";
 import { DatabaseModule } from "./database/database.module";
-import { ConfigController } from "./config/config.controller";
+import { ConfigController } from "./route-controllers/config/config.controller";
```
Comment (fwpushan) [inline sources/packages/api/src/app.module.ts:5]: Add config controller export in the route-controller/index.ts and then import the controller in line 7

**Review by fwpushan — COMMENTED** [review]
Excellent job. Just only small comment. Is it possible to add a basic test suit to validate this flow. or we can add technical debt. I think an e2e test suit is required for this.

File: sources/packages/api/src/app.module.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/app.module.ts]: Goot catch! Code changed to import the ConfigController and ConfigService using their respective index.ts files.

**Comment by andrewsignori-aot** [comment]
> Excellent job. Just only small comment. Is it possible to add a basic test suit to validate this flow. or we can add technical debt. I think an e2e test suit is required for this.

Tech debt created: Secure API endpoint - E2E tests#174

File: sources/packages/api/src/auth/auth.module.ts [hunk 2]
```diff
@@ -1,37 +1,29 @@
...
     },
     {
       provide: APP_GUARD,
-      useClass: RoleGuard,
-    },*/
+      useClass: RolesGuard,
+    },
   ],
 })
 export class AuthModule {}
```
Comment (2011sandeepraj) [inline sources/packages/api/src/auth/auth.module.ts:29]: Remove the parseAuthorizationHeader methods

File: sources/packages/api/src/route-controllers/config/config.controller.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { Controller, Get } from "@nestjs/common";
+import { Public } from "src/auth/decorators/public.decorator";
+import { ConfigService } from "src/services/config/config.service";
+import { IConfig } from "./models/get-config.dto";
```
Comment (2011sandeepraj) [inline sources/packages/api/src/route-controllers/config/config.controller.ts:4]: Let's standardize import paths. Use Relative Paths

**Review by 2011sandeepraj — COMMENTED** [review]
Some Minor COmments

File: sources/packages/api/src/auth/auth.module.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/auth/auth.module.ts:29]: Done, good catch 👍

File: sources/packages/api/src/route-controllers/config/config.controller.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/config/config.controller.ts:4]: Done!

**Comment by andrewsignori-aot** [comment]
> Some Minor COmments

I went through the comments and are all done.

**Review by fwpushan — APPROVED** [review]

**Pull request merged**